be/src/olap/rowset/alpha_rowset_reader.cpp

```diff
@@ -136,57 +127,70 @@ int64_t AlphaRowsetReader::filtered_rows() {
+        OLAPStatus status = _next_block_for_column_data(pos, &_row_blocks[pos]);
+        if (status == OLAP_ERR_DATA_EOF) {
+            continue;
+        } else if (status != OLAP_SUCCESS) {
+            return status;
+        } else {
+            (*block) = _row_blocks[pos];
+            return OLAP_SUCCESS;
         }
-        if (_key_range_indices[pos] >= _key_range_size) {
-            *row_block = nullptr;
-            return OLAP_ERR_DATA_EOF;
+    }
+    if (pos == _row_blocks.size()) {
+        *block = nullptr;
+        return OLAP_ERR_DATA_EOF;
+    }
+
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus AlphaRowsetReader::_merge_block(RowBlock** block) {
+    // Row among different segment groups may overlap with each other.
+    // Iterate all row_blocks to fetch min row each round.
+    OLAPStatus status = OLAP_SUCCESS;
+    _read_block->clear();
+    size_t num_rows_in_block = 0;
+    while (_read_block->pos() < _num_rows_per_row_block) {
+        RowCursor* row_cursor = nullptr;
+        status = _next_row_for_singleton_rowset(&row_cursor);
+        if (status == OLAP_ERR_DATA_EOF && _read_block->pos() > 0) {
+            status = OLAP_SUCCESS;
+            break;
+        } else if (status != OLAP_SUCCESS) {
+            return status;
         }
-        return OLAP_SUCCESS;
+        _read_block->get_row(_read_block->pos(), _dst_cursor);
+        _dst_cursor->copy(*row_cursor, _read_block->mem_pool());
+        _read_block->pos_inc();
+        num_rows_in_block++;
     }
+    _read_block->set_pos(0);
+    _read_block->set_limit(num_rows_in_block);
+    _read_block->finalize(num_rows_in_block);
+    *block = _read_block.get();
     return status;
 }
 
-OLAPStatus AlphaRowsetReader::_get_next_row_for_singleton_rowset(RowCursor** row) {
+OLAPStatus AlphaRowsetReader::_next_row_for_singleton_rowset(RowCursor** row) {
     RowCursor* min_row = nullptr;
     int min_index = -1;
     for (int i = 0; i < _row_blocks.size(); i++) {
-        if (_row_blocks[i] == nullptr) {
-            continue;
-        }
-        RowCursor* current_row = _row_cursors[i];
-        if (!_row_blocks[i]->has_remaining()) {
-            OLAPStatus status = _get_next_block(i, &_row_blocks[i]);
-            if (status != OLAP_SUCCESS) {
-                LOG(WARNING) << "_get_next_block failed, status:" << status;
+        if (_row_blocks[i] == nullptr || !_row_blocks[i]->has_remaining()) {
+            OLAPStatus status = _next_block_for_column_data(i, &_row_blocks[i]);
+            if (status == OLAP_ERR_DATA_EOF) {
+                continue;
+            } else if (status != OLAP_SUCCESS) {
+                LOG(INFO) << "read next row of singleton rowset failed:" << status;
                 return status;
             }
         }
         size_t pos = _row_blocks[i]->pos();
+        RowCursor* current_row = _row_cursors[i];
         _row_blocks[i]->get_row(pos, current_row);
         if (min_row == nullptr || min_row->cmp(*current_row) <  0) {
```

Comment:
```suggestion
        if (min_row == nullptr || min_row->cmp(*current_row) >  0) {
```
I think this should be large than

Reply:
OK